### PR TITLE
feat: token-protected content playback

### LINKED ISSUE #28: Token protected playback
**As a user**, I would like to be able to play Akamai token-protected content offered by **SRG SSR**.

## Criteria 

The content is playable on
- different types of device such as `Desktop`, `Tablet`, `Mobile`
- different OS such as `Window`, `Apple`, `Linux`
- different `browsers`

Supported content includes
- `VoD`s
- `Live` streams with or without `DVR`

### PR DESCRIPTION
**Pull request by amtins**
## Description

Resolves #28 and allows content protected by an Akamai token to be played.

## Changes made

- add `AkamaiTokenService` to hangle token generation
- update `SrgSsr` to handle token-protected resources
- update the demo
- update package.json `version`
- add test coverage

### CODE REVIEW

File: src/utils/AkamaiTokenService.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  static aclPath(streamUrl) {
+    let acl;
+
+    if (streamUrl.href.includes('/hls/playingLive/')) {
+      acl = '/hls/playingLive/*';
+    } else {
+      const path = streamUrl.pathname;
+
+      acl = `${path.substring(0, path.lastIndexOf('/') + 1)}*`;
+    }
```
Comment (defagos) [inline src/utils/AkamaiTokenService.js:23]: Interesting, we have a [different way](https://github.com/SRGSSR/pillarbox-apple/blob/main/Sources/CoreBusiness/ResourceLoading/AkamaiResourceLoaderDelegate.swift#L23-L29) of building the token. Ours seems a bit more generic, this is probably something we should discuss.

File: test/utils/AkamaiTokenService.spec.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
+import AkamaiTokenService from '../../src/utils/AkamaiTokenService.js';
+import streamTypeList from '../__mocks__/streamTypeList.js';
+
+window.fetch = jest.fn((source) => {
+  return Promise.resolve({
+    ok: source.includes('succes'),
```
Comment (defagos) [inline test/utils/AkamaiTokenService.spec.js:6]:
```suggestion
    ok: source.includes('success'),
```

Maybe a typo here?

File: test/utils/AkamaiTokenService.spec.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    status: 503,
+    statusText: 'Service Unavailable',
+    json: () => {
+      if (source.includes('succes')) {
```
Comment (defagos) [inline test/utils/AkamaiTokenService.spec.js:10]:
```suggestion
      if (source.includes('success')) {
```

Maybe a typo here?

File: test/utils/AkamaiTokenService.spec.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+   * tokenize ******************************************************************
+   *****************************************************************************
+   */
+  describe('tokenize', () => {
+    it('should return a tokenized source', async () => {
+      const spyOnObject = jest.spyOn(Object, 'assign');
+
+      const result = await AkamaiTokenService.tokenize(
+        { url: 'https://mock-token.ch/with/a/long/path/manifest.m3u8' },
+        'https://succes.token.server.ch'
```
Comment (defagos) [inline test/utils/AkamaiTokenService.spec.js:102]:
```suggestion
        'https://success.token.server.ch'
```

Maybe a typo here?

File: test/utils/AkamaiTokenService.spec.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    it('should match the pattern', () => {
+      const url = new URL(
+        'https://fake.url/hls/playingLive/with/a/very/long/path'
+      );
+
+      expect(AkamaiTokenService.aclPath(url)).toBe('/hls/playingLive/*');
+    });
+    it('should not match the pattern', () => {
+      const url = new URL(
+        'https://fake.url/cotent/not/playingLive/with/a/very/long/path'
```
Comment (defagos) [inline test/utils/AkamaiTokenService.spec.js:37]:
```suggestion
        'https://fake.url/conent/not/playingLive/with/a/very/long/path'
```

Typo.

File: test/utils/AkamaiTokenService.spec.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+      );
+
+      expect(AkamaiTokenService.aclPath(url)).toBe(
+        '/cotent/not/playingLive/with/a/very/long/*'
```
Comment (defagos) [inline test/utils/AkamaiTokenService.spec.js:41]:
```suggestion
        '/content/not/playingLive/with/a/very/long/*'
```

Typo.

File: test/utils/AkamaiTokenService.spec.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+   *****************************************************************************
+   */
+  describe('tokenizeSources', () => {
+    it('should return a tokenized sources', async () => {
+      const result = await AkamaiTokenService.tokenizeSources(
+        [
+          { url: 'https://mock-token.ch/with/a/long/path/manifest.m3u8' },
+          { url: 'https://mock-token.ch/with/a/long/path/manifest2.m3u8' }
+        ],
+        'https://succes.token.server.ch'
```
Comment (defagos) [inline test/utils/AkamaiTokenService.spec.js:133]:
```suggestion
        'https://success.token.server.ch'
```

Typo.

File: src/utils/AkamaiTokenService.js (hunk 1 above)
Comment (amtins) [reply src/utils/AkamaiTokenService.js:23]: @defagos I prefer your solution, it looks cleaner! Let's discuss it on Monday. 

If I remember correctly, I asked the android devs how they generated their URLs and I applied the same rule to Letterbox, so I imported the same logic.

FYI, here's what Android does today in [pillarbox](https://github.com/SRGSSR/pillarbox-android/blob/5008a1015023cbff075b3427e5c1859a84e4f528/pillarbox-core-business/src/main/java/ch/srgssr/pillarbox/core/business/akamai/AkamaiTokenProvider.kt#L41-L46)

**Comment by amtins** [comment]
@defagos thanks for the review 👍🏽

File: src/utils/AkamaiTokenService.js (hunk 1 above)
Comment (defagos) [reply src/utils/AkamaiTokenService.js:23]: Great, I guess you are both wrong then 😉. Let us discuss this implementation on Monday then.

Comment (amtins) [reply src/utils/AkamaiTokenService.js]: After the discussion we add, I removed the check for the `/hls/playingLive/` which might be some sort of legacy code. Thank you @defagos 👍🏽

**Review by defagos — APPROVED** [review]

**Pull request merged**